zebra/zebra_northbound.c

```diff
@@ -0,0 +1,2212 @@
+}
+
+/*
+ * XPath:
+ * /frr-route-map:lib/route-map/entry/match-condition/frr-zebra:source-protocol
+ */
+static int lib_route_map_entry_match_condition_source_protocol_modify(
+	enum nb_event event, const struct lyd_node *dnode,
+	union nb_resource *resource)
+{
+	struct routemap_hook_context *rhc;
+	const char *type;
+	int rv;
+
+	switch (event) {
+	case NB_EV_VALIDATE:
+		type = yang_dnode_get_string(dnode, NULL);
+		if (proto_name2num(type) == -1) {
+			zlog_warn("%s: invalid protocol: %s", __func__, type);
+			return NB_ERR_VALIDATION;
+		}
+		return NB_OK;
+	case NB_EV_PREPARE:
+	case NB_EV_ABORT:
+		return NB_OK;
+	case NB_EV_APPLY:
+		/* NOTHING */
+		break;
+	}
+
+	/* Add configuration. */
+	rhc = nb_running_get_entry(dnode, NULL, true);
+	type = yang_dnode_get_string(dnode, NULL);
+
+	/* Set destroy information. */
+	rhc->rhc_mhook = generic_match_delete;
+	rhc->rhc_rule = "source-protocol";
+	rhc->rhc_event = RMAP_EVENT_MATCH_DELETED;
+
+	rv = generic_match_add(NULL, rhc->rhc_rmi, "source-protocol", type,
+			       RMAP_EVENT_MATCH_ADDED);
+	if (rv != CMD_SUCCESS) {
+		rhc->rhc_mhook = NULL;
+		return NB_ERR_INCONSISTENCY;
+	}
+
+	return NB_OK;
+}
+
+static int lib_route_map_entry_match_condition_source_protocol_destroy(
+	enum nb_event event, const struct lyd_node *dnode)
+{
+	return lib_route_map_entry_match_destroy(event, dnode);
+}
+
+/*
+ * XPath:
+ * /frr-route-map:lib/route-map/entry/match-condition/frr-zebra:source-instance
+ */
+static int lib_route_map_entry_match_condition_source_instance_modify(
+	enum nb_event event, const struct lyd_node *dnode,
+	union nb_resource *resource)
+{
+	struct routemap_hook_context *rhc;
+	const char *type;
+	int rv;
+
+	if (event != NB_EV_APPLY)
+		return NB_OK;
+
+	/* Add configuration. */
+	rhc = nb_running_get_entry(dnode, NULL, true);
+	type = yang_dnode_get_string(dnode, NULL);
+
+	/* Set destroy information. */
+	rhc->rhc_mhook = generic_match_delete;
+	rhc->rhc_rule = "source-instance";
+	rhc->rhc_event = RMAP_EVENT_MATCH_DELETED;
+
+	rv = generic_match_add(NULL, rhc->rhc_rmi, "source-instance", type,
+			       RMAP_EVENT_MATCH_ADDED);
+	if (rv != CMD_SUCCESS) {
+		rhc->rhc_mhook = NULL;
+		return NB_ERR_INCONSISTENCY;
+	}
+
+	return NB_OK;
+}
+
+static int lib_route_map_entry_match_condition_source_instance_destroy(
+	enum nb_event event, const struct lyd_node *dnode)
+{
+	return lib_route_map_entry_match_destroy(event, dnode);
+}
+
+/*
+ * XPath: /frr-route-map:lib/route-map/entry/set-action/frr-zebra:source-v4
+ */
+static int
+lib_route_map_entry_set_action_source_v4_modify(enum nb_event event,
+						const struct lyd_node *dnode,
+						union nb_resource *resource)
+{
+	struct routemap_hook_context *rhc;
+	struct interface *pif = NULL;
+	const char *source;
+	struct vrf *vrf;
+	struct prefix p;
+	int rv;
+
+	switch (event) {
+	case NB_EV_VALIDATE:
+		memset(&p, 0, sizeof(p));
+		yang_dnode_get_ipv4p(&p, dnode, NULL);
+		if (zebra_check_addr(&p) == 0) {
```

Comment:
Do we need this check? In yang we are using ipv4-address for this leaf so check should already happen.

Reply:
That check does more than that and I want to keep the previous behavior.

Reply:
I think these checks can be added in Yang itself with range as well.

Reply:
```c
int zebra_check_addr(const struct prefix *p)
{
        if (p->family == AF_INET) {
                uint32_t addr;

                addr = p->u.prefix4.s_addr;
                addr = ntohl(addr);

                if (IPV4_NET127(addr) || IN_CLASSD(addr)
                    || IPV4_LINKLOCAL(addr))
                        return 0;
        }
        if (p->family == AF_INET6) {
                if (IN6_IS_ADDR_LOOPBACK(&p->u.prefix6))
                        return 0;
                if (IN6_IS_ADDR_LINKLOCAL(&p->u.prefix6))
                        return 0;
        }
        return 1;
}
```
This is quite a lot to check using YANG, I personally think code-level validation is fine in this case. If you think otherwise, maybe propose something different on a separate PR later? I'd like to unlock this one since it's opened for several months already.

Reply:
Sure. Having this check does not harm but we can revisit this with separate PR.